src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs

```diff
@@ -516,6 +528,23 @@ public static Vector3 Transform(Vector3 position, Matrix4x4 matrix)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Vector3 Transform(Vector3 value, Quaternion rotation)
         {
+            if (Vector128.IsHardwareAccelerated)
+            {
+                var vVector = value.AsVector128();
+                var rVector = Unsafe.BitCast<Quaternion, Vector128<float>>(rotation);
+
+                // v + 2 * (q x (q.W  * v + q x v)
+                return (vVector + Vector128.Create(2f) * Cross(rVector, Vector128.Shuffle(rVector, Vector128.Create(3, 3, 3, 3)) * vVector + Cross(rVector, vVector))).AsVector3();
+
+                static Vector128<float> Cross(Vector128<float> v1, Vector128<float> v2)
+                {
+                    return (Vector128.Shuffle(v1, Vector128.Create(1, 2, 0, 3)) *
+                            Vector128.Shuffle(v2, Vector128.Create(2, 0, 1, 3))) -
+                           (Vector128.Shuffle(v1, Vector128.Create(2, 0, 1, 3)) *
+                            Vector128.Shuffle(v2, Vector128.Create(1, 2, 0, 3)));
+                }
+            }
```

Comment:
I'm not immediately following where this implementation is from...

The standard naive algorithm is typically broken down to something like:
```csharp
return ((Quaternion.Conjugate(rotation) * new Quaternion(value, 0.0f)) * rotation).AsVector128().AsVector3();
```

(noting that `Quaternion.operator *(Quaternion, Quaterion)` is not currently marked as aggressively inline)

Reply:
I got it from a private C library I wrote a decade ago which quotes a dead link.
The closest I could find on the internet after a quick search is this [forum post](https://fgiesen.wordpress.com/2019/02/09/rotating-a-single-vector-using-a-quaternion/) which gets 95% of the way there and just misses the last 'simplification for readability' step.

Reply:
Can we update this to follow the "standard" algorithm I gave above.

I expect we'll get better support for things like constant folding, more consistent performance across a range of hardware, and we can accurately quote the root implementation to a known source (which avoids any potential licensing issues): https://github.com/microsoft/DirectXMath/blob/main/Inc/DirectXMathVector.inl#L10307-L10317

IIRC, `glm` uses a similar algorithm for it's implementation

Reply:
The naive implementation is even slower then the dotnet8 implementation, even with the dotnet9 improvements to quaternion multiplication. It will also always be slower then the reduced version because it contains significantly more math operation.

I also don't see any potential licensing problem. I wrote the c# code. Btw it looks nothing like c code I based it on which I also wrote. Both version are based on math that cannot be copyrighted and that has been floating around the internet for at least decade at this point.

To reduce any variation between different archs the implementation could be changed to 
```
return (vVector + 2 * Cross(rVector, Vector128.Create(rotation.W) * vVector + Cross(rVector, vVector))).AsVector3();
```
which would give the JIT a bit more leeway, but I doubt it makes any differens.

Reply:
> The naive implementation is even slower then the dotnet8 implementation, even with the dotnet9 improvements to quaternion multiplication

This is in part because `quaternion multiplication` is not being inlined currently. But, also because it's doing some "unnecessary" work, as you indicated. But that can be accounted for as well.

> I also don't see any potential licensing problem. I wrote the c# code. Btw it looks nothing like c code I based it on which I also wrote. Both version are based on math that cannot be copyrighted and that has been floating around the internet for at least decade at this point.

In general we require correct attribution to exist and based on the statement given above, the code (even if you authored it) was itself based on some internet blog/forum/site/etc post that can no longer be referenced (due to a dead link). Such a post may have itself may have been based on something else which overall makes it very difficult for us to use "safely" as the chain of citations is broken.

The actual requirements for attribution, whether or not something can be copyrighted, and the like can get quite complex. When there isn't a clear chain anymore, then it requires additional effort on our part, potentially even requiring checks with legal, so an actual lawyer can make the determination on whether or not it is safe to do.

.NET is a huge open source project depending on by millions. We must do the due diligence and ensure that all the boxes are checked.

Reply:
What would make this safe is if you took the naive algorithm and provided a comment inline with the code giving how you arrive at a new algorithm via simplification.

i.e. Document the algorithm is `return ((Quaternion.Conjugate(rotation) * new Quaternion(value, 0.0f)) * rotation).AsVector128().AsVector3();`, as taken from the DirectX Math source material and then show the work for how simplification is arrived at

Then covering that quaternion multiplication is:
```csharp
Vector3 av = value1.AsVector3();
Vector3 bv = value2.AsVector3();

Vector3 c = Vector3.Cross(av, bv)
float dot = Vector3.Dot(av, bv);

return (value1 * value2.W)
     + (value2 * value1.W).WithElement(3, 0)
     + new Quaternion(c, dot);
```

and quaternion conjugate is:
```csharp
return new Quaternion(
    -value.X,
    -value.Y,
    -value.Z,
     value.W
);
```

Therefore, the operation is:
```csharp
Vector3 av1 = -(rotation.AsVector3());
Vector3 bv1 = value;

Vector3 c1 = Vector3.Cross(av1, bv1)
float dot1 = Vector3.Dot(av1, bv1);

Quaternion tmp1 = (rotation * 0)
                + new Quaternion(value * rotation.W, 0)
                + new Quaternion(c1, dot1);

Vector3 av2 = tmp1.AsVector3();
Vector3 bv2 = rotation.AsVector3();

Vector3 c2 = Vector3.Cross(av2, bv2)
float dot2 = Vector3.Dot(av2, bv2);

Quaternion tmp2 = (tmp1 * rotation.W)
                + (rotation * tmp1.W).WithElement(3, 0)
                + new Quaternion(c2, dot2);

return tmp2.AsVector3();
```

This simplifies trivially a bit:
```csharp
Vector3 av1 = -(rotation.AsVector3());
Vector3 bv1 = value;

Vector3 c1 = Vector3.Cross(av1, bv1)
float dot1 = Vector3.Dot(av1, bv1);

Quaternion tmp1 = new Quaternion(value * rotation.W, 0)
                + new Quaternion(c1, dot1);

Vector3 av2 = tmp1.AsVector3();
Vector3 bv2 = rotation.AsVector3();

Vector3 c2 = Vector3.Cross(av2, bv2)
float dot2 = Vector3.Dot(av2, bv2);

Quaternion tmp2 = (tmp1 * rotation.W)
                + (rotation * dot1).WithElement(3, 0)
                + new Quaternion(c2, dot2);

return tmp2.AsVector3();
```

Then, since we know we don't need the W component, it simplifies more:
```csharp
Vector3 rot = rotation.AsVector3();

Vector3 av1 = -rot;
Vector3 bv1 = value;

Vector3 c1 = Vector3.Cross(av1, bv1)
float dot1 = Vector3.Dot(av1, bv1);

Vector3 av2 = (bv1 * rotation.W) + c1;
Vector3 bv2 = rot;

Vector3 c2 = Vector3.Cross(av2, bv2);

return (av2 * rotation.W) + (bv2 * dot1) + c2;
```

Finally with some removing of unnecessary locals:
```csharp
Vector3 rot = rotation.AsVector3();
Vector3 av1 = -rot;

Vector3 av2 = (value * rotation.W) + Vector3.Cross(av1, value);
return (av2 * rotation.W) + (rot * Vector3.Dot(av1, value)) + Vector3.Cross(av2, rot);
```

------

This example of a simplification ends up faster than the current scalar code. It's not quite as fast as the code you've provided, but it follows an appropriate and documentable process.

It's also much more trivial to see that it returns the same result. The blog post you linked above (https://fgiesen.wordpress.com/2019/02/09/rotating-a-single-vector-using-a-quaternion/) walks through a simplification, but not in terms of code but rather in terms of bog standard math. It takes several assumptions which aren't strictly safe and which could (although I've not done the work to validate if that's actually the case) cause deviations due to `computer math` not being the same as `real math` -- Which is to say, `a + (b + c)` is not guaranteed to return the same as `(a + b) + c` and not all identities hold the same, particularly due to the existence of concepts like `-0`, `Infinity`, `NaN`, `Subnormals`, and the like.

Reply:
> What is the expected behavior around NaN/Infinity? Aren't most math function garbage in/garbage out?

`NaN` and `Infinity` are not garbage, they are well-defined values that exist and are important for many transcendental functions as well as for doing operations involving `Complex` numbers.

APIs have input domains and values outside that domain should return `NaN`. Values inside the input domain need to follow the logically correct behavior for such an input, which is that the operation is done as if to infinite precision and unbounded range and then rounded to the nearest representable result.
-- As per the IEEE 754 spec, in practice we tend to allow a little bit of variance away from the infinitely precise result for performance reasons, but that tends to be well documented/understood (such as we do for `lerp`).

> libeigen also appears to be using a very similar version for at least 16 years https://gitlab.com/libeigen/eigen/-/blame/8de4d92b70753e81420d2ff387f9596eb423b2de/Eigen/src/Geometry/Quaternion.h?page=1#L185

That code is also LGPL which brings about implications if that's cited as a reference.

>  So the math has heritage, but the code does not.

The heritage of the math and/or code all matters. We have to show our work and how we arrive at the simplified algorithm or cite the reference that did the work for us. If that reference is licensed (as many published math papers and most code are), then we need to ensure that we are inline with those license terms.

Unless we can actually show the work on how you arrive at simplifying from the standard algorithm to the simplified version or reference a well known code example under a compatible license, then we should default to using one of the well known implementations that we can cite, such as what I had given above.

Reply:
> That code is also LGPL which brings about implications if that's cited as a reference.

> The heritage of the math and/or code all matters. We have to show our work and how we arrive at the simplified algorithm or cite the reference that did the work for us. If that reference is licensed (as many published math papers and most code are), then we need to ensure that we are inline with those license terms.

> Unless we can actually show the work on how you arrive at simplifying from the standard algorithm to the simplified version or reference a well known code example under a compatible license, then we should default to using one of the well known implementations that we can cite, such as what I had given above.

Then how about [ceres-solver (apache)](https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/rotation.h#L722) or [android (also apache)](https://android.googlesource.com/platform//frameworks/native/+/1cd746b6abd542010741c216593beb3b6c56dd66/libs/math/include/math/TQuatHelpers.h#108) or [glfw (zlib)](https://github.com/glfw/glfw/blob/master/deps/linmath.h#L483) or [khronos forum post from 2004 (who knows what license khronos forums have)](https://community.khronos.org/t/quaternions-and-tangent-space/39831/6) or [the unreadable 2006 wayback machine copy of the post my c code was based on (GNU Free Documentation License)](https://web.archive.org/web/20061118031122/http://gns.wikia.com/wiki/Quaternions).

My point is that is a pretty standard way of multiplying a vector by a quaternion that is used all over the place and who exactly was the first person to do some algebra to reduce the number of operations is probably lost to time.

If the requirement is to write some algebra as a comment I can do that, but if the requirement is to trace every piece of math back to some 1960' compsci student I might as well give up.

Reply:
The requirement is to simply document the chain of trust. It is no different than writing a paper where you need to provide citations and as such can be achieved by showing the algebra giving how the simplification is come to -or- can be from giving the citation to another trusted source.

------------------------

In this particular case, however, there is an additional consideration that I tried to describe as well which is that just because something can be simplified mathematically does not necessarily mean it can be simplified programmatically. 

Mathematically `x + 2 + 3` is the same as `x + 5`. Programmatically they can return different results and thus `(x + 2) + 3` is distinct from `x + (2 + 3)`.

Which is to say, it is also important to consider correctness in addition to performance and while many engines may arrive at the mathematical simplification that can lead to potentially skewed results in various edge cases and it may be overall undesirable.

Even the simplification I gave above is only safe because we throw away the `W` component from `rotation`. Normalizing `x * 0` to `0` isn't normally safe because `-0 * 0 == -0`. So we need to track that `(rotation * 0)` could produce `-0` and then consider that `-0 + x == x` so it is safe. However, `0 + x` may not always produce `x` since `-0 + 0 == 0`. So we have to track potential values and how they would propagate into `tmp1` to see that it's effectively throw away and therefore should be safe to simplify.

This is notably why the more de-facto implementations as provided by `DirectX Math` and `GLM` (OpenGL Math Library) instead do something more akin to what I had arrived at, because while the algorithm you gave is faster I don't believe it to be as correct in all scenarios.

Reply:
`-0 * 0 == -0` is nice and all, but the method makes no guaranties on how the vector is rotated. The documentation only claims that 

> Transforms a vector by the specified Quaternion rotation value.

Many people would assume that means Q\*vQ, but converting the quaternion into a 3x3 rotation matrix and multiplying that or disassembling the quaternion into angles and applying Rodrigues' three times would be equally valid. If your code relies on the believe that the internal implementation of Vector3.Transform is Q\*vQ your code is wrong and you should just use Q\*vQ directly.

Reply:
That's not how APIs in the .NET BCL work. While we don't necessarily guarantee a strict behavior, we do typically try to aim for correctness and there is a spec described correct behavior that is derived from the naive mathematical algorithm.

These APIs are all built on `float` which is an IEEE 754 defined type. The strict behavior, defined by IEEE 754, is that each operation should be computed as if to infinite precision and unbounded range and then rounded to the nearest representable result. We do strictly follow this behavior (on scalar types like `float`, `double`, or `half`) for IEEE 754 required operations (such as `+`, `*`, `-`, `/`, `sqrt`, etc). However for optional operations (such as `sin`, `cos`, `hypot`, etc) we instead tend to only guarantee special boundary conditions (such as `+/-0`, `nan`, `+/-infinity`, `+/-1`, values outside the input domain, and sometimes including special considerations like evens vs odds or midpoints, etc).

`Vector3.Transform(Vector3 value, Quaternion rotation)` falls outside both of these in that it is an API exposed on a custom type but we try to treat it in the same category as we do "optional operations" and therefore we want to ensure boundary conditions remain well-handled and deterministic across platform. We then try to get the rest of the result accurate to within a reasonable ULP (units in last place, where the spec compliant result is accurate to within 0.5 ULP). That is to say, we try to balance the performance vs accuracy characteristic to ensure accurate results can be returned while still allow some level of non-determinism across different computers so that we can take advantage of common optimizations that don't impact boundary values so that results can also be returned promptly.

Not considering these edge cases is what leads to problems and causes jitter, bugs, and other fundamental problems. Perf is not the end all here and it is instead up to the end user to write their own algorithm if they want to deviate from the correct defaults. It can also lead to subtle breaks over time if we just change the behavior without considering what the prior behavior and edge case handling had been.

Reply:
What makes the result of Q\*vQ any more correct that any of the other ways you could use to rotate a vector by the rotation saved in a quaternion? Rodrigues' formula is actually older than Hamilton's, but it's not like any of this is a universal law written in the stars. This entire conversation feels pointless so I will stop wasting everyone's time, close this pull request and just use a private library.

Reply:
From a mathematical perspective they all produce the "same" result. The consideration is that they are not the same from a programmatic perspective due to additional considerations.

The intent was to push this PR in a direction where we could still improve the performance but without changing certain invariants that are extremely important to maintain. That requires more in depth consideration and documentation to show how those requirements are being met.

The intent was not to waste time or get into an argument over what is right vs wrong. The contribution was appreciated and was otherwise nearly in the right shape to be taken.

Reply:
I know and appreciate that you weren't trying to waste any time, but I just don't have the spare time to have long debates about computational vs mathematical correctness right now, so I think it's just easier if I leave it at that so that someone else can make a pull request if they run into the same bottlenecks.

Reply:
👍, are you fine with me cherry-picking your commits and otherwise fixing the last couple bits of feedback so this can get merged and you can still get credit for the overall contribution?

Reply:
Sure